module-system/sov-modules-macros/src/cli_parser.rs

```diff
@@ -14,90 +16,416 @@ impl CliParserMacro {
+                }
+            }
+
+            let enum_defn = quote! {
+                #[derive(::clap::Parser)]
+                pub enum #item_with_named_fields_ident #generics {
+                    #(#variants_with_named_fields,)*
+                }
+            };
+
+            let from_body = quote! {
+                match item {
+                    #(#convert_cases)*
+                }
+            };
+            (enum_defn, from_body)
+        }
+        Data::Struct(s) => {
+            let named_fields = build_named_fields_with_docs(&s.fields);
+            let field_names = named_fields.iter().map(|f| &f.ident).collect::<Vec<_>>();
+            let conversion_logic = match s.fields {
+                Fields::Named(_) => quote! {{
+                        let #item_with_named_fields_ident { #(#field_names),* } = item;
+                        #item_name{#(#field_names),*}
+                }},
+                Fields::Unnamed(_) => {
+                    quote! {
+                            let #item_with_named_fields_ident { #(#field_names),* } = item;
+                            #item_name(#(#field_names),*)
+                    }
+                }
+                Fields::Unit => quote! { #item_name },
+            };
+
+            let struct_defn = quote! {
+                #[derive(::clap::Parser)]
+                pub struct #item_with_named_fields_ident #generics {
+                    #(#named_fields),*
+                }
+            };
+            (struct_defn, conversion_logic)
+        }
+        Data::Union(_) => {
+            return Err(syn::Error::new_spanned(
+                ast,
+                "Unions are not supported as CLI wallet args",
+            ))
+        }
+    };
+
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let clap_type = if is_generic {
+        quote! { #item_with_named_fields_ident #ty_generics }
+    } else {
+        quote! { #item_with_named_fields_ident }
+    };
+
+    let expanded = quote! {
+        // Create a new data type which matches the original, but with all fields named.
+        // This is the type that Clap will parse the CLI args into
+        #named_type_defn
+
+        // Define a `From` implementation which converts from the named fields version to the original version
+        impl #impl_generics From<#clap_type> for #item_name #ty_generics #where_clause {
+            fn from(item: #item_with_named_fields_ident #ty_generics) -> Self {
+                #conversion_logic
+            }
+        }
+
+        // Implement the `CliWalletArg` trait for the original type. This is what allows the original type to be used as a CLI arg.
+        impl #impl_generics sov_modules_api::CliWalletArg for #item_name #ty_generics #where_clause {
+            type CliStringRepr = #clap_type;
+        }
+    };
+    Ok(expanded.into())
+}
+
+/// A struct or enum field with a name and type
+pub struct NamedFieldWithDocs {
```

Comment:
We have a`StructNamedField` type in `common.rs`  which looks similar. Is it possible to extend `StructNamedField` and use it instead of `NamedFieldWithDocs`? Can be done in another PR.

Reply:
Migrated!